proto/symge/symge-demo.cpp

```diff
@@ -0,0 +1,387 @@
+// bloody CMake
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+#include <cassert>
+#include <cstring>
+#include <iomanip>
+#include <iostream>
+#include <queue>
+#include <random>
+#include <sstream>
+#include <string>
+#include <unordered_set>
+#include <utility>
+
+#include <util/optional.hpp>
+
+#include "symbolic.hpp"
+#include "msparse.hpp"
+
+using namespace sym;
+
+// Identifier name picking helper routines
+
+void join_impl(std::ostream& ss, const std::string& sep) {}
+
+template <typename Head, typename... Args>
+void join_impl(std::ostream& ss, const std::string& sep, Head&& head, Args&&... tail) {
+    if (sizeof...(tail)==0)
+        ss << std::forward<Head>(head);
+    else
+        join_impl(ss << std::forward<Head>(head) << sep, sep, std::forward<Args>(tail)...);
+}
+
+template <typename... Args>
+std::string join(const std::string& sep, Args&&... items) {
+    std::stringstream ss;
+    join_impl(ss, sep, std::forward<Args>(items)...);
+    return ss.str();
+}
+
```

Comment:
comments please. Just explain what it does. This might be useful for naming temporary variables in modcc?